Source/ObjectiveDropboxOfficial/Shared/Handwritten/OAuth/DBOAuthManager.m

```diff
@@ -365,13 +372,18 @@ - (void)extractFromUrl:(NSURL *)url completion:(DBOAuthCompletion)completion {
 - (void)finishPkceOAuthWithAuthCode:(NSString *)authCode
                        codeVerifier:(NSString *)codeVerifier
                          completion:(DBOAuthCompletion)completion {
+  [_sharedApplication presentLoading];
```

Comment:
This is weird. I think the delegation should be inverted. Meaning `DBOAuthManager` should actually be emitting these "request started" and "request ended" events out instead of it asking an instance of `DBSharedApplication` to `presentLoading` or `dismissLoading`. The oauth manager should not decide what should be shown when and where. A listening object should decide that.

I would say lets have a protocol named:

`DBOAuthManagerDelegate` that has two methods:

`oauthManagerDidBeginRefresh:(DBOAuthManager *)oauthManager`

and

`oauthManagerDidCompleteRefresh:(DBOAuthManager *)oauthManager`

I would then set the delegate to be `DBMobileSharedApplication` that listens in on these events and does the appropriate thing.

Reply:
I think both ways are fine. The main reason I used it this way instead of delegation method is because `DBSharedApplication` already has a few "UI presenting" methods that OAuthManager calls directly, e.g. `presentErrorMessage:title:`.  So it might be reasonable to keep the usages consistent. what do you think?

Reply:
I actually feel that something like the oauth manager is core business logic and should have no knowledge about UI and when things should appear/disappear.

Reply:
I totally agree with your from the clean architecture perspective. however this code isn't written in a way that clear boundaries are drawn for each layer. The oauth manager right now is really a mix of interactor + repository and some control over presentation too. I think it would make more sense we refactor the class as an tech improvement in a long run, and for this PR, I don't feel good enough to do partial optimization/refectory and I'd rather keep consistent with existing code.